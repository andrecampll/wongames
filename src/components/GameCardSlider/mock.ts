export default [
  {
    title: 'Population Zero',
    developer: 'Rockstar Games',
    image: 'https://source.unsplash.com/user/willianjusten/300x140',
    price: 235,
    promotionalPrice: 215,
  },
  {
    title: 'Population Zero',
    developer: 'Rockstar Games',
    image: 'https://source.unsplash.com/user/willianjusten/300x141',
    price: 235,
    promotionalPrice: 215,
  },
  {
    title: 'Population Zero',
    developer: 'Rockstar Games',
    image: 'https://source.unsplash.com/user/willianjusten/300x142',
    price: 235,
    promotionalPrice: 215,
  },
  {
    title: 'Population Zero',
    developer: 'Rockstar Games',
    image: 'https://source.unsplash.com/user/willianjusten/300x143',
    price: 235,
    promotionalPrice: 215,
  },
  {
    title: 'Population Zero',
    developer: 'Rockstar Games',
    image: 'https://source.unsplash.com/user/willianjusten/300x144',
    price: 235,
    promotionalPrice: 215,
  },
  {
    title: 'Population Zero',
    developer: 'Rockstar Games',
    image: 'https://source.unsplash.com/user/willianjusten/300x145',
    price: 235,
    promotionalPrice: 215,
  },
];
