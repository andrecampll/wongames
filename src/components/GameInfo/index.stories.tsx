import { Story, Meta } from '@storybook/react/types-6-0';
import GameInfo, { GameInfoProps } from '.';
import { CartContextDTO } from '../../hooks/cart/useCart';
import mockGame from './mock';

export default {
  title: 'Game/GameInfo',
  component: GameInfo,
  parameters: {
    backgrounds: {
      default: 'dark',
    },
  },
  args: mockGame,
} as Meta;

export const Default: Story<GameInfoProps> = args => (
  <div style={{ maxWidth: '144rem', margin: 'auto', padding: '1.5rem' }}>
    <GameInfo {...args} />
  </div>
);

export const IsInCart: Story<GameInfoProps & CartContextDTO> = args => (
  <div style={{ maxWidth: '144rem', margin: 'auto', padding: '1.5rem' }}>
    <GameInfo {...args} />
  </div>
);

IsInCart.args = {
  isInCart: () => true,
};
