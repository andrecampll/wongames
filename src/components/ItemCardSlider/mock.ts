export default [
  {
    id: '1',
    name: 'Camisa',
    slug: 'camisa-0',
    image: 'https://source.unsplash.com/user/willianjusten/300x140',
    price: 235,
    promotionalPrice: 215,
  },
  {
    id: '2',
    name: 'Camisa',
    slug: 'camisa-0',
    image: 'https://source.unsplash.com/user/willianjusten/300x141',
    price: 235,
    promotionalPrice: 215,
  },
  {
    id: '3',
    name: 'Camisa',
    slug: 'camisa-0',
    image: 'https://source.unsplash.com/user/willianjusten/300x142',
    price: 235,
    promotionalPrice: 215,
  },
  {
    id: '4',
    name: 'Camisa',
    slug: 'camisa-0',
    image: 'https://source.unsplash.com/user/willianjusten/300x143',
    price: 235,
    promotionalPrice: 215,
  },
  {
    id: '5',
    name: 'Camisa',
    slug: 'camisa-0',
    image: 'https://source.unsplash.com/user/willianjusten/300x144',
    price: 235,
    promotionalPrice: 215,
  },
  {
    id: '6',
    name: 'Camisa',
    slug: 'camisa-0',
    image: 'https://source.unsplash.com/user/willianjusten/300x145',
    price: 235,
    promotionalPrice: 215,
  },
];
