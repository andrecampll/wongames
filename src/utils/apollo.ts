import { useMemo } from 'react';
import { setContext } from '@apollo/client/link/context';
import {
  ApolloClient,
  HttpLink,
  NormalizedCache,
  NormalizedCacheObject,
} from '@apollo/client';
import { Session } from 'next-auth';
import apolloCache from './apolloCache';

let apolloClient: ApolloClient<NormalizedCacheObject | null>;

function createApolloClient(session?: Session | null) {
  const httpLink = new HttpLink({
    uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql`,
  });

  const authLink = setContext((_, { headers, session: clientSession }) => {
    const jwt = session?.jwt || clientSession?.jwt || '';
    const authorization = jwt ? `Bearer ${jwt}` : '';

    return { headers: { ...headers, authorization } };
  });

  return new ApolloClient({
    ssrMode: typeof window === 'undefined',
    link: authLink.concat(httpLink),
    cache: apolloCache,
  });
}

export function initializeApollo(
  initialState = null,
  session?: Session | null,
) {
  const apolloClientGlobal = apolloClient ?? createApolloClient(session);

  if (initialState) {
    apolloClientGlobal.cache.restore(
      initialState as NormalizedCacheObject & NormalizedCache,
    );
  }

  if (typeof window === 'undefined') return apolloClientGlobal;

  apolloClient = apolloClient ?? apolloClientGlobal;

  return apolloClient;
}

export function useApollo(initialState = null, session?: Session) {
  const store = useMemo(() => initializeApollo(initialState, session), [
    initialState,
    session,
  ]);

  return store;
}
