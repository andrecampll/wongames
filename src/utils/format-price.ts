export function formatPrice(price: number | bigint): string {
  return new Intl.NumberFormat('en', {
    style: 'currency',
    currency: 'USD',
  }).format(price);
}
