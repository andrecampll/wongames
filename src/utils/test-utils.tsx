import { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';

import {
  WishlistContext,
  WishlistContextDefaultValues,
  WishlistContextDTO,
} from '../hooks/wishlist/useWishlist';

import theme from '../styles/theme';
import {
  CartContext,
  CartContextDTO,
  CartContextDefaultValues,
} from '../hooks/cart/useCart';

type CustomRenderProps = {
  cartProviderProps?: CartContextDTO;
  wishlistProviderProps?: WishlistContextDTO;
} & Omit<RenderOptions, 'queries'>;

const customRender = (
  ui: ReactElement,
  {
    cartProviderProps = CartContextDefaultValues,
    wishlistProviderProps = WishlistContextDefaultValues,
    ...renderOptions
  }: CustomRenderProps = {},
) =>
  render(
    <ThemeProvider theme={theme}>
      <CartContext.Provider value={cartProviderProps}>
        <WishlistContext.Provider value={wishlistProviderProps}>
          {ui}
        </WishlistContext.Provider>
      </CartContext.Provider>
    </ThemeProvider>,
    renderOptions,
  );

export * from '@testing-library/react';
export { customRender as render };
