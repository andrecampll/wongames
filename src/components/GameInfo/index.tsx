import { formatPrice } from '../../utils/format-price';

import Heading from '../Heading';
import Ribbon from '../Ribbon';
import { CartButton } from '../CartButton';
import { WishlistButton } from '../WishlistButton';

import { Wrapper, Description, ButtonsWrapper } from './styles';

export type GameInfoProps = {
  id: string;
  title: string;
  description: string;
  price: number;
};

const GameInfo = ({ id, title, description, price }: GameInfoProps) => (
  <Wrapper>
    <Heading color="black" lineBottom>
      {title}
    </Heading>

    <Ribbon color="secondary">{formatPrice(price)}</Ribbon>

    <Description>{description}</Description>

    <ButtonsWrapper>
      <CartButton id={id} size="large" hasText />
      <WishlistButton id={id} hasText size="large" />
    </ButtonsWrapper>
  </Wrapper>
);

export default GameInfo;
