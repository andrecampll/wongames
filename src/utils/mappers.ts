import { QueryGames_games } from '../graphql/generated/QueryGames';
import {
  QueryHome_banners,
  QueryHome_sections_freeGames_highlight,
} from '../graphql/generated/QueryHome';
import { QueryWishlist_wishlists_games } from '../graphql/generated/QueryWishlist';

import { formatPrice } from './format-price';

export const bannerMapper = (banners: QueryHome_banners[]) => {
  return banners.map(({ image, title, subtitle, button, ribbon }) => ({
    image: `http://localhost:1337${image?.url}`,
    title,
    subtitle,
    buttonLabel: button?.label,
    buttonLink: button?.link,
    ...(ribbon && {
      ribbon: ribbon.text,
      ribbonColor: ribbon.color,
      ribbonSize: ribbon.size,
    }),
  }));
};

export const gamesMapper = (games: QueryGames_games[] | null | undefined) => {
  return games
    ? games.map(({ id, name, slug, developers, cover, price }) => ({
        id,
        title: name,
        slug,
        developer: developers[0].name,
        image: cover?.url
          ? `http://localhost:1337${cover?.url}`
          : 'https://i.stack.imgur.com/y9DpT.jpg',
        price,
      }))
    : [];
};

export const highlightMapper = (
  highlight: QueryHome_sections_freeGames_highlight | null | undefined,
) => {
  return highlight
    ? {
        title: highlight.title,
        subtitle: highlight.subtitle,
        backgroundImage: `http://localhost:1337${highlight.background.url}`,
        floatImage: `http://localhost:1337${highlight.floatImage.url}`,
        buttonLabel: highlight.buttonLabel,
        buttonLink: highlight.buttonLink,
        alignment: highlight.alignment,
      }
    : {};
};

export const cartMapper = (
  games:
    | QueryGames_games[]
    | QueryWishlist_wishlists_games[]
    | null
    | undefined,
) => {
  return games
    ? games.map(game => ({
        id: game.id,
        img: `http://localhost:1337${game.cover?.url}`,
        title: game.name,
        price: formatPrice(game.price),
      }))
    : [];
};
