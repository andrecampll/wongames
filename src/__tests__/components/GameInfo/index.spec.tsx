import '../../../../.jest/session.mock';
import { screen } from '@testing-library/react';
import { render } from '../../../utils/test-utils';
import GameInfo from '../../../components/GameInfo';

const props = {
  id: '1',
  title: 'My Game',
  description: 'Game description',
  price: 210,
};

describe('GameInfo', () => {
  it('should be able to render game information', () => {
    const { container } = render(<GameInfo {...props} />);

    expect(
      screen.getByRole('heading', {
        name: /my game/i,
      }),
    ).toBeInTheDocument();

    expect(screen.getByText(/game description/i)).toBeInTheDocument();
    expect(screen.getByText(/\$210\.00/)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be able to render', () => {
    render(<GameInfo {...props} />);

    expect(
      screen.getByRole('button', {
        name: /add to cart/i,
      }),
    ).toBeInTheDocument();

    expect(
      screen.getByRole('button', {
        name: /add to wishlist/i,
      }),
    ).toBeInTheDocument();
  });
});
