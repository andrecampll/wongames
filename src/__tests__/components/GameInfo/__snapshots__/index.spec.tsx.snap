// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameInfo should be able to render game information 1`] = `
<div
  class="sc-eCssSg dHsiHY"
>
  <h2
    class="sc-bdfBwQ gvBnTC"
    color="black"
  >
    My Game
  </h2>
  <div
    class="sc-gsTCUz fWqWz"
    color="secondary"
  >
    $210.00
  </div>
  <p
    class="sc-jSgupP kZrrCD"
  >
    Game description
  </p>
  <div
    class="sc-gKsewC coMbzQ"
  >
    <button
      class="sc-dlfnbm jZeNUn"
    >
      <svg
        aria-hidden="true"
        aria-label="Add to cart"
        class="StyledIconBase-ea9ulj-0 cKCFNq"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-8.9-5h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4l-3.87 7H8.53L4.27 2H1v2h2l3.6 7.59-1.35 2.44C4.52 15.37 5.48 17 7 17h12v-2H7l1.1-2z"
        />
      </svg>
      <span>
        Add to cart
      </span>
    </button>
    <button
      class="sc-dlfnbm hRsQbH"
    >
      <svg
        aria-hidden="true"
        aria-label="Add to Wishlist"
        class="StyledIconBase-ea9ulj-0 cKCFNq"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
        />
      </svg>
      <span>
        Add to Wishlist
      </span>
    </button>
  </div>
</div>
`;
